openai/client.py

```diff
@@ -0,0 +1,392 @@
+        self._credential = credential
+        self._cached_token = None
+
+    def get_token(self) -> str:
+        if (
+            self._cached_token is None
+            or (self._cached_token.expires_on - time.time()) < 300
+        ):
+            self._cached_token = self._credential.get_token(
+                "https://cognitiveservices.azure.com/.default"
+            )
+
+        return self._cached_token.token
+
+
+class ApiKeyAuth:
+    """Authentication using an API key"""
+
+    def __repr__(self):
+        return f"ApiKeyAuth(api_key=<redacted>)"
+    
+    def __init__(self, key: str = ""):
+        """Create a new ApiKeyAuth instance. If no key is passed, it will use ~openai.api_key"""
+        self.key = key or openai.api_key
+
+    def get_token(self) -> str:
+        """Get the API key"""
+        return self.key
+
+
+Backends = typing.Literal["azure", "openai", ""]
+
+
+class OpenAIClient:
+    def __init__(
+        self,
+        *,
+        api_base: str = "",
+        auth: typing.Union[str, ApiKeyAuth, AzureTokenAuth] = "",
+        api_version: str = "",
+        backend: Backends = "",
+    ):
+        """Create a new OpenAI client.
+           
+           :param: api_base: The base URL for the API. If not specified, based on ~opeanai.api_base
+           :param: auth: The authentication method or key to use. If the string value "azuredefault" is passed, it will use ~azure.identity.DefaultAzureCredential
+           :param: api_version: The API version to use. If not specified, based on ~openai.api_version or ~openai.client.LATEST_AZURE_API_VERSION.
+           :param backend: One of 'azure' or 'openai'. If not specified, inferred from the auth method or ~openai.api_type
+        """
+
+        #
+        # This code is a bit messy, but it's because we want to hide the messiness from the caller.
+        #
+        if auth == "azuredefault":
+            self.auth = AzureTokenAuth()
+        elif isinstance(auth, str):
+            self.auth = ApiKeyAuth(auth or openai.api_key)
+        else:
+            self.auth = auth
+
+        # Pick up api type from parameter or environment
+        backend = backend or (
+            "azure" if openai.api_type in ("azure", "azure_ad") or isinstance(auth, AzureTokenAuth) else "openai"
+        )
+
+        self.backend = backend
+
+        if backend == "azure":
+            self.api_version = (
+                api_version or openai.api_version or LATEST_AZURE_API_VERSION
+            )
+            if isinstance(self.auth, AzureTokenAuth):
+                self.api_type = "azure_ad"
+            else:
+                self.api_type = "azure"
+        elif backend == "openai":
+            self.api_type = "open_ai"
+            self.api_version = api_version or openai.api_version
+        else:
+            raise ValueError(
+                f'Unknown `backend` {backend} - expected one of "azure" or "openai"'
+            )
+        if not api_base and backend == 'azure':
```

Comment:
What if I want to pass my azure api_base via environment variable?

Reply:
Yeah, that should work. Will fix.

Reply:
Ah, I now remember - the problem is that the default is openai, and for azure you always need to have overridden the value. I'll put a check to make sure the `api_base` isn't "https://api.openai.com/v1". I'll have to go back and re-untangle this mess I think.